File: marcus_crosby/project04/Primecountbench.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class Primecountbench {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java Primecountbench <input-file> [thread-count]");
            return;
        }

        Path inputPath = Path.of(args[0]);
        int threadCount = args.length >= 2 ? parseThreadCount(args[1]) : Runtime.getRuntime().availableProcessors();

        try {
            List<Long> numbers = readNumbers(inputPath);
            if (numbers.isEmpty()) {
                System.out.println("No numbers found in file: " + inputPath);
                return;
            }

            long singleStart = System.nanoTime();
            long singlePrimeCount = countPrimeSingleThread(numbers);
            long singleElapsedNanos = System.nanoTime() - singleStart;

            long multiStart = System.nanoTime();
            long multiPrimeCount = countPrimeMultiThread(numbers, threadCount);
            long multiElapsedNanos = System.nanoTime() - multiStart;

            System.out.println("Input file: " + inputPath);
            System.out.println("Total numbers parsed: " + numbers.size());
            System.out.println();
            System.out.println("Single-thread:");
            System.out.println("  Prime count: " + singlePrimeCount);
            System.out.printf("  Elapsed time: %.3f ms%n", nanosToMillis(singleElapsedNanos));
            System.out.println();
            System.out.println("Multi-thread (" + threadCount + " threads):");
            System.out.println("  Prime count: " + multiPrimeCount);
            System.out.printf("  Elapsed time: %.3f ms%n", nanosToMillis(multiElapsedNanos));

            if (singlePrimeCount != multiPrimeCount) {
                System.out.println();
                System.out.println("Warning: counts do not match between modes.");
            }
        } catch (IOException e) {
            System.err.println("Failed to read file: " + e.getMessage());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.err.println("Execution interrupted.");
        } catch (ExecutionException e) {
            System.err.println("Worker task failed: " + e.getCause());
        }
    }

    private static int parseThreadCount(String text) {
        try {
            int parsed = Integer.parseInt(text);
            return Math.max(1, parsed);
        } catch (NumberFormatException e) {
            return Runtime.getRuntime().availableProcessors();
        }
    }

    private static List<Long> readNumbers(Path inputPath) throws IOException {
        String content = Files.readString(inputPath);
        String[] tokens = content.split("[^0-9-]+");
        List<Long> numbers = new ArrayList<>();

        for (String token : tokens) {
            if (token.isBlank() || "-".equals(token)) {
                continue;
            }
            try {
                numbers.add(Long.parseLong(token));
            } catch (NumberFormatException ignored) {
                // Skip values outside long range
            }
        }
        return numbers;
    }

    private static long countPrimeSingleThread(List<Long> numbers) {
        long count = 0;
        for (long n : numbers) {
            if (isPrime(n)) {
                count++;
            }
        }
        return count;
    }

    private static long countPrimeMultiThread(List<Long> numbers, int threadCount)
            throws InterruptedException, ExecutionException {
        int size = numbers.size();
        int workers = Math.min(threadCount, size);
        int chunkSize = (size + workers - 1) / workers;
        ExecutorService executor = Executors.newFixedThreadPool(workers);

        try {
            List<Future<Long>> futures = new ArrayList<>();
            for (int i = 0; i < size; i += chunkSize) {
                int from = i;
                int to = Math.min(i + chunkSize, size);
                Callable<Long> task = () -> {
                    long localCount = 0;
                    for (int j = from; j < to; j++) {
                        if (isPrime(numbers.get(j))) {
                            localCount++;
                        }
                    }
                    return localCount;
                };
                futures.add(executor.submit(task));
            }

            long total = 0;
            for (Future<Long> future : futures) {
                total += future.get();
            }
            return total;
        } finally {
            executor.shutdown();
        }
    }

    private static boolean isPrime(long n) {
        if (n < 2) {
            return false;
        }
        if (n == 2 || n == 3) {
            return true;
        }
        if (n % 2 == 0 || n % 3 == 0) {
            return false;
        }

        for (long i = 5; i * i <= n; i += 6) {
            if (n % i == 0 || n % (i + 2) == 0) {
                return false;
            }
        }
        return true;
    }

    private static double nanosToMillis(long nanos) {
        return nanos / 1_000_000.0;
    }
}

File: marcus_crosby/project04/Primecountbench.kt
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors

fun main(args: Array<String>) {
    if (args.isEmpty()) {
        println("Usage: kotlin Primecountbench.kt <input-file> [thread-count]")
        return
    }

    val inputPath = Path.of(args[0])
    val threadCount = if (args.size >= 2) parseThreadCount(args[1]) else Runtime.getRuntime().availableProcessors()

    try {
        val numbers = readNumbers(inputPath)
        if (numbers.isEmpty()) {
            println("No numbers found in file: $inputPath")
            return
        }

        val singleStart = System.nanoTime()
        val singlePrimeCount = countPrimeSingleThread(numbers)
        val singleElapsedNanos = System.nanoTime() - singleStart

        val multiStart = System.nanoTime()
        val multiPrimeCount = countPrimeMultiThread(numbers, threadCount)
        val multiElapsedNanos = System.nanoTime() - multiStart

        println("Input file: $inputPath")
        println("Total numbers parsed: ${numbers.size}")
        println()
        println("Single-thread:")
        println("  Prime count: $singlePrimeCount")
        println("  Elapsed time: %.3f ms".format(nanosToMillis(singleElapsedNanos)))
        println()
        println("Multi-thread ($threadCount threads):")
        println("  Prime count: $multiPrimeCount")
        println("  Elapsed time: %.3f ms".format(nanosToMillis(multiElapsedNanos)))

        if (singlePrimeCount != multiPrimeCount) {
            println()
            println("Warning: counts do not match between modes.")
        }
    } catch (e: IOException) {
        System.err.println("Failed to read file: ${e.message}")
    } catch (e: InterruptedException) {
        Thread.currentThread().interrupt()
        System.err.println("Execution interrupted.")
    } catch (e: ExecutionException) {
        System.err.println("Worker task failed: ${e.cause}")
    }
}

private fun parseThreadCount(text: String): Int =
    text.toIntOrNull()?.coerceAtLeast(1) ?: Runtime.getRuntime().availableProcessors()

private fun readNumbers(inputPath: Path): List<Long> {
    val content = Files.readString(inputPath)
    val tokens = content.split(Regex("[^0-9-]+"))
    val numbers = mutableListOf<Long>()

    for (token in tokens) {
        if (token.isBlank() || token == "-") {
            continue
        }
        token.toLongOrNull()?.let { numbers.add(it) }
    }

    return numbers
}

private fun countPrimeSingleThread(numbers: List<Long>): Long {
    var count = 0L
    for (n in numbers) {
        if (isPrime(n)) {
            count++
        }
    }
    return count
}

private fun countPrimeMultiThread(numbers: List<Long>, threadCount: Int): Long {
    val size = numbers.size
    val workers = minOf(threadCount, size)
    val chunkSize = (size + workers - 1) / workers
    val executor = Executors.newFixedThreadPool(workers)

    try {
        val futures = mutableListOf<java.util.concurrent.Future<Long>>()

        var i = 0
        while (i < size) {
            val from = i
            val to = minOf(i + chunkSize, size)
            val task = Callable<Long> {
                var localCount = 0L
                for (j in from until to) {
                    if (isPrime(numbers[j])) {
                        localCount++
                    }
                }
                localCount
            }
            futures.add(executor.submit(task))
            i += chunkSize
        }

        var total = 0L
        for (future in futures) {
            total += future.get()
        }
        return total
    } finally {
        executor.shutdown()
    }
}

private fun isPrime(n: Long): Boolean {
    if (n < 2) return false
    if (n == 2L || n == 3L) return true
    if (n % 2L == 0L || n % 3L == 0L) return false

    var i = 5L
    while (i * i <= n) {
        if (n % i == 0L || n % (i + 2L) == 0L) {
            return false
        }
        i += 6L
    }

    return true
}

private fun nanosToMillis(nanos: Long): Double = nanos / 1_000_000.0
